scipy/optimize/_canonical_constraint.py

```diff
@@ -0,0 +1,480 @@
+from __future__ import division, print_function, absolute_import
+import numpy as np
+import scipy.sparse as spc
+from ._constraints import (NonlinearConstraint,
+                           LinearConstraint,
+                           BoxConstraint)
+
+__all__ = ['CanonicalConstraint',
+           'to_canonical',
+           'lagrangian_hessian',
+           'empty_canonical_constraint']
+
+
+class CanonicalConstraint:
+    """Canonical constraint
+
+    Constraint of the form:
+        c_ineq <= 0
+        c_eq = 0
+    for:
+        c_ineq, c_eq = constr(x)
+    """
+    def __init__(self, n_vars, n_ineq, n_eq,
+                 constr, jac, hess, sparse_jacobian,
+                 enforce_feasibility,
+                 x0, c_ineq0, c_eq0, J_ineq0, J_eq0):
+        # Dimensions
+        self.n_vars = n_vars
+        self.n_ineq = n_ineq
+        self.n_eq = n_eq
+        # Objective function and constraints
+        self.constr = constr
+        self.jac = jac
+        self.hess = hess
+        # Use sparse jacobian flag
+        self.sparse_jacobian = sparse_jacobian
+        # Enforce feasibility for CanonicalConstraint. Should
+        # be a list of booleans (and never a single boolean value,
+        # as it is allowed for Box, Linear and Nonlinear constraints).
+        self.enforce_feasibility = enforce_feasibility
+        # Initial Values
+        self.x0 = x0
+        self.c_ineq0 = c_ineq0
+        self.c_eq0 = c_eq0
+        self.J_ineq0 = J_ineq0
+        self.J_eq0 = J_eq0
+
+
+def to_canonical(constraints):
+    """Convert constraints or list of constraints to canonical format."""
+    # Put ``constraints`` in list format whe needed
+    if isinstance(constraints, (NonlinearConstraint,
+                                LinearConstraint,
+                                BoxConstraint,
+                                CanonicalConstraint)):
+        constraints = [constraints]
+    if isinstance(constraints, (list, tuple, np.array)):
+        # Converts all constraints to canonical format
+        constraints_list = []
+        for c in constraints:
+            if isinstance(c, CanonicalConstraint):
+                constraints_list += [c]
```

Comment:
Looks like the case for `append` method.

Reply:
ok